src/reify.cpp

```diff
@@ -0,0 +1,265 @@
+
+Reifier::~Reifier() noexcept = default;
+
+template <typename... T>
+void Reifier::printFact(const char* name, const T&... args) {
+    out_ << name << "(";
+    printComma(out_, args...);
+    out_ << ").\n";
+}
+
+template <typename... T>
+void Reifier::printStepFact(const char* name, const T&... args) {
+    if (reifyStep_) {
+        printFact(name, args..., step_);
+    } else {
+        printFact(name, args...);
+    }
+}
+
+template <typename M, typename T>
+auto Reifier::tuple(M& map, const char* name, const T& args) -> size_t {
+    return tuple(map, name, toVec(args));
+}
+
+template <typename M, typename T>
+auto Reifier::tuple(M& map, const char* name, std::vector<T>&& args) -> size_t {
+    auto ret = map.emplace(std::move(args), map.size());
+    if (ret.second) {
+        printStepFact(name, ret.first->second);
+        for (const auto& x : ret.first->first) {
+            printStepFact(name, ret.first->second, x);
+        }
+    }
+    return ret.first->second;
+}
+
+template <typename M, typename T>
+auto Reifier::orderedTuple(M& map, const char* name, const T& args) -> size_t {
+    return orderedTuple(map, name, toVec(args));
+}
+
+template <typename M, typename T>
+auto Reifier::orderedTuple(M& map, const char* name, std::vector<T>&& args) -> size_t {
+    auto ret = map.emplace(std::move(args), map.size());
+    if (ret.second) {
+        printStepFact(name, ret.first->second);
+        int arg = 0;
+        for (const auto& x : ret.first->first) {
+            printStepFact(name, ret.first->second, arg, x);
+            ++arg;
+        }
+    }
+    return ret.first->second;
+}
+
+auto Reifier::theoryTuple(IdSpan args) -> size_t {
+    return orderedTuple(stepData_.theoryTuples, "theory_tuple", args);
+}
+
+auto Reifier::theoryElementTuple(IdSpan args) -> size_t {
+    return tuple(stepData_.theoryElementTuples, "theory_element_tuple", args);
+}
+
+auto Reifier::litTuple(LitSpan args) -> size_t {
+    return tuple(stepData_.litTuples, "literal_tuple", args);
+}
+
+auto Reifier::weightLitTuple(WeightLitSpan args) -> size_t {
+    WLVec lits;
+    lits.reserve(args.size());
+    for (const auto& x : args) {
+        lits.emplace_back(x.lit, x.weight);
+    }
+    return tuple(stepData_.weightLitTuples, "weighted_literal_tuple", std::move(lits));
+}
+
+auto Reifier::atomTuple(AtomSpan args) -> size_t {
+    return tuple(stepData_.atomTuples, "atom_tuple", args);
+}
+
+auto Reifier::addNode(Atom_t atom) -> Graph::Node& {
+    auto& node = stepData_.nodes_[atom];
+    if (!node) {
+        node = &stepData_.graph_.insertNode(atom);
+    }
+    return *node;
+}
+
+void Reifier::initProgram(bool incremental) {
+    if (incremental) {
+        printFact("tag", "incremental");
+    }
+}
+
+void Reifier::beginStep() {}
+
+void Reifier::rule(HeadType ht, AtomSpan head, LitSpan body) {
+    const char* h = ht == HeadType::disjunctive ? "disjunction" : "choice";
+    std::ostringstream hss, bss;
```

Comment:
I tend to make such things members to cut down the number of allocations.

Reply:
If string streams are used only once, one can use `std::move(stream).str()` to move the string out without extra allocation. If the just a view is required, stream.view() can be used.

Reply:
...or use types/printValue overlods.